### PR TITLE
Make (some) loaders serializable

### PR DESCRIPTION
**Pull request by logan-markewich**
# Description

This PR adds some serialization to some of the more popular loaders in llama-index.

A follow-up would be converting some llama-hub loaders to be the same, once the new base class is merged in.

In addition, there is a new class that can help represent loaders and the inputs intended for them.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Added new unit/integration tests
- [x] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama_index/readers/base.py [hunk 1]
```diff
@@ -1,10 +1,14 @@
 """Base reader class."""
 from abc import abstractmethod
-from typing import Any, List
+from typing import Any, Dict, List
 
-from llama_index.bridge.langchain import Document as LCDocument
+try:
+    from pydantic.v1 import Field
+except ImportError:
+    from pydantic import Field
```
Comment (Disiok) [inline llama_index/readers/base.py:8]: Made a PR to consolidate all the try/except logic into a single bridge file: https://github.com/jerryjliu/llama_index/pull/7502

File: llama_index/readers/base.py [hunk 2]
```diff
@@ -18,3 +22,30 @@ def load_langchain_documents(self, **load_kwargs: Any) -> List[LCDocument]:
...
+class PydanticBaseReader(BaseReader, BaseComponent):
+    """Serialiable Data Loader with Pydatnic."""
+
+    is_remote: bool = False
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class LoaderConfig(BaseComponent):
```
Comment (Disiok) [inline llama_index/readers/base.py:36]: Sometimes I find it pretty confusing that we call it loader sometimes, reader other times, and connectors remaining times.

File: llama_index/readers/base.py [hunk 3]
```diff
@@ -18,3 +22,30 @@ def load_langchain_documents(self, **load_kwargs: Any) -> List[LCDocument]:
...
+    """Represents a loader and it's input arguments."""
+
+    loader: BaseReader = Field(..., description="Loader to use.")
```
Comment (Disiok) [inline llama_index/readers/base.py:39]: putting notes here from offline discussion:
1. loader ctor (& pydantic attributes) configures "how to connect to API"
2. args/kwargs configures "what data to load"

Comment (Disiok) [reply llama_index/readers/base.py:39]: 2 corresponds to `def load_data` calls.

File: llama_index/readers/web.py [hunk 4]
```diff
@@ -23,6 +27,11 @@ class SimpleWebPageReader(BaseReader):
             Default is None.
     """
 
+    is_remote: bool = True
+    html_to_text: bool
+
+    _metadata_fn: Optional[Callable[[str], Dict]] = PrivateAttr()
```
Comment (Disiok) [inline llama_index/readers/web.py:33]: Double checking: right now serializing/deserializing would forget any custom metadata fn

File: llama_index/readers/base.py (hunk 1 above)
Comment (logan-markewich) [reply llama_index/readers/base.py:8]: Thank you! I've been meaning to do that lol

File: llama_index/readers/web.py (hunk 4 above)
Comment (logan-markewich) [reply llama_index/readers/web.py:33]: Yea, some loaders have configs that aren't serializable

The beautiful soup loader is another example

File: llama_index/readers/wikipedia.py [hunk 5]
```diff
@@ -1,18 +1,20 @@
...
     """Wikipedia reader.
 
     Reads a page.
 
     """
 
-    def __init__(self) -> None:
+    is_remote: bool = True
+
+    def __init__(self, pages: Optional[List[str]] = None) -> None:
```
Comment (Disiok) [inline llama_index/readers/wikipedia.py:17]: I think this should be reverted?

File: llama_index/readers/twitter.py [hunk 6]
```diff
@@ -28,6 +32,19 @@ def __init__(
         """Initialize with parameters."""
         self.bearer_token = bearer_token
         self.num_tweets = num_tweets
```
Comment (Disiok) [inline llama_index/readers/twitter.py:34]: very nit: don't need to set attributes anymore?

File: llama_index/readers/twitter.py [hunk 7]
```diff
@@ -28,6 +32,19 @@ def __init__(
...
+        super().__init__(
+            is_remote=True,
```
Comment (Disiok) [inline llama_index/readers/twitter.py:36]: very nit: already configured at the class level?

**Review by Disiok — APPROVED** [review]
LGTM aside from very minor comments.

File: llama_index/readers/base.py (hunk 2 above)
Comment (logan-markewich) [reply llama_index/readers/base.py:36]: Ranemed to `ReaderConfig` :)

File: llama_index/readers/twitter.py (hunk 6 above)
Comment (logan-markewich) [reply llama_index/readers/twitter.py:34]: Actually it was there before I started lol

**Pull request merged**